### ISSUE TITLE
[RFC] Make the root component configurable in app.json instead of being hardcoded to "main" in package.json

### ISSUE THREAD

**Issue by Arrow7000** [expo/expo-cli#1487]
This would make it much simpler to support monorepos with two or more apps from a shared codebase.

If you want to export multiple apps you currently need two app.json files and you can start each one by running expo start --config path/to/app.json. This is really easy.

The problem is that you can't run two apps in parallel because the root component has to be set in the repo's package.json's "main" property. This means manually changing the "main" to point to the other app's root component every single time you are switching apps.

The only solution currently is having multiple package.json files, which is a maintenance nightmare, especially if you want to share a lot of code between apps.

TL;DR putting the "main" property in app.json instead of package.json means that all the information about each app is set in the app.json and it enables running multiple apps in parallel and easier deployment.
<hr/>

[Link to the original feature request on canny.io](https://expo.canny.io/feature-requests/p/make-the-root-component-configurable-in-appjson-instead-of-being-hardcoded-to-ma)

**Comment by Arrow7000** [comment]
After some digging around in the source - and testing this on a multi-app project -  it looks as if this feature already exists... by using `"entryPoint"` in app.json
https://github.com/expo/expo-cli/blob/d582e54f93e9045522c1852c10f36c3a393de915/packages/config/src/paths/paths.ts#L66
In which case, why is this not documented anywhere??

**Comment by brentvatne** [comment]
hey @Arrow7000 - i think it is documented: https://docs.expo.io/versions/latest/workflow/configuration/#entrypoint

it's not documented more prominently because this isn't really a commonly used feature and i'm not sure where else we would put it.

it's not clear to me why you would run the same app simultaneously from the same directory. this feels like an edge case that could be worked around by an individual developer as needed. can you convince me that we should support running multiple instances of the same app from one directory?

**Comment by Arrow7000** [comment]
@brentvatne thanks for replying.

I think the right place for documenting entryPoint would be where it tells you how to change the location of the root component: https://docs.expo.io/versions/latest/sdk/register-root-component/#what-if-i-want-to-name-my
It currently only mentions package.json's main even though you could also use app.json's entryPoint.

My use case isn't for running two instances of the same app, it's for running two different apps in a single repo. E.g. if you have a monorepo with a single package.json and some shared components but want to publish two separate apps:

```
repo/
├── package.json
├── shared-components/
│   ├── Header.tsx
│   └── Body.tsx
├── appA/
│   ├── AppA.tsx
│   └── appA.json
└── appB/
     ├── AppB.tsx
     └── appB.json
```

**Comment by brentvatne** [comment]
interesting, couldn't you run `expo start` in both `appA` and `appB`?

**Comment by brentvatne** [comment]
we have a similar kind of set up in the `expo/expo` repo - https://github.com/expo/expo

you can run each of the apps in the `apps` directory at the same time, in addition to the one in the `home` directory - https://github.com/expo/expo/tree/master/apps / https://github.com/expo/expo/tree/master/home

**Comment by Arrow7000** [comment]
When I tried that on my (Windows 10) machine it tries to start both apps on the same port so only one of them works. This may be an issue with Metro bundler though

**Comment by brentvatne** [comment]
gotcha. I would suggest using yarn workspaces and have a look at how we configured things in this repo: https://github.com/expo/expo/tree/master/packages/expo-yarn-workspaces

if you want to open another rfc to make a case for why you should be able to do `expo start` multiple times in the same project on different ports then please do that :) this title of this particular issue seems like it is already resolved via the entryPoint property that you found, so i'll close
